File: backend/app/schemas/knowledge.py
"""Knowledge item + document/processing schemas."""

from __future__ import annotations

import datetime
import uuid

from app.schemas.common import CamelModel


class AIEnrichmentSchema(CamelModel):
    """The per-memory AI facet (mirrors ``AIEnrichment`` in the frontend).

    Stored on ``KnowledgeItem.ai`` as a JSON dict with snake_case keys;
    ``CamelModel`` (``populate_by_name``) validates those and serializes the
    camelCase shape (``keyMoments``, ``actionItems``) the client renders.
    """

    title: str | None = None
    summary: str | None = None
    # Free-form on the read side so unexpected stored values never 500 a list;
    # the enricher only ever writes positive|neutral|negative|mixed.
    mood: str | None = None
    topics: list[str] = []
    key_moments: list[str] = []
    action_items: list[str] = []


class KnowledgeItemSchema(CamelModel):
    id: uuid.UUID
    workspace_id: uuid.UUID
    collection_id: uuid.UUID | None = None
    user_id: uuid.UUID
    type: str
    title: str
    content: str | None = None
    excerpt: str | None = None
    source_url: str | None = None
    preview_url: str | None = None
    thumbnail_url: str | None = None
    file_url: str | None = None
    status: str
    embedding_status: str
    chunk_count: int
    page_count: int
    tags: list[str]
    ai: AIEnrichmentSchema | None = None
    is_favorite: bool
    occurred_at: datetime.datetime | None = None
    last_accessed_at: datetime.datetime | None = None
    created_at: datetime.datetime
    updated_at: datetime.datetime


class KnowledgeItemCreate(CamelModel):
    workspace_id: uuid.UUID
    type: str = "note"
    title: str
    content: str | None = None
    excerpt: str | None = None
    source_url: str | None = None
    tags: list[str] = []
    collection_id: uuid.UUID | None = None


class KnowledgeItemUpdate(CamelModel):
    title: str | None = None
    content: str | None = None
    excerpt: str | None = None
    tags: list[str] | None = None
    is_favorite: bool | None = None
    collection_id: uuid.UUID | None = None


class ProcessingJobSchema(CamelModel):
    id: uuid.UUID
    knowledge_item_id: uuid.UUID | None = None
    status: str
    stage: str
    progress: float
    error: str | None = None
    retry_count: int
    max_retries: int


class UploadJobSchema(CamelModel):
    id: uuid.UUID
    filename: str
    content_type: str | None = None
    size_bytes: int
    checksum: str | None = None
    storage_key: str | None = None
    public_url: str | None = None
    status: str


class UploadResponse(CamelModel):
    item: KnowledgeItemSchema
    upload_job: UploadJobSchema
    processing_job: ProcessingJobSchema


class UrlUploadRequest(CamelModel):
    source_url: str
    workspace_id: uuid.UUID | None = None


class ProcessingStatus(CamelModel):
    job_id: uuid.UUID
    status: str
    stage: str
    progress: float
    error: str | None = None
    retry_count: int
    stages: list[dict] = []
    preview: str | None = None

File: src/types/knowledge.ts
import type { HasTimestamps, ID, ISODateString } from "./common";
import type { KnowledgePipeline, PipelineStage } from "./pipeline";

/**
 * What kind of memory/knowledge a captured item represents. The first six are
 * the original set; the rest extend coverage to every supported source. An
 * image memory uses `photo`, the existing visual composition.
 */
export type KnowledgeType =
  | "note"
  | "voice"
  | "photo"
  | "document"
  | "link"
  | "idea"
  | "pdf"
  | "video"
  | "website"
  | "bookmark"
  | "chat";

/**
 * State of a memory's vector embeddings — what AI Chat needs before it can
 * retrieve from this item. Mirrors the backend embedding job; the client never
 * computes embeddings, it only reflects their status.
 */
export type EmbeddingStatus = "pending" | "embedding" | "ready" | "failed";

/**
 * RAG-facing facets of a memory: everything AI Chat needs to retrieve from and
 * cite this item. Optional on the model — the backend fills it during ingestion;
 * until then `deriveRag()` provides a best-effort view from the pipeline.
 */
export interface KnowledgeRag {
  /** Stable id the vector store and citations reference. */
  documentId: ID;
  embeddingStatus: EmbeddingStatus;
  /** Number of chunks the document was split into. */
  chunkCount: number;
  /** Page count, for paginated sources (pdf / document). */
  pageCount?: number;
  /** Current ingestion stage — mirrors `KnowledgePipeline.stage`. */
  processingStage: PipelineStage;
}

/** AI-detected sentiment of an item. */
export type Mood = "positive" | "neutral" | "negative" | "mixed";

/**
 * Coarse processing status used for list badges and filtering. Derived from the
 * detailed {@link KnowledgePipeline} via `getProcessingStatus()` — never stored
 * separately, so the two can't drift.
 */
export type ProcessingStatus = "uploading" | "processing" | "ready" | "failed";

export interface Person {
  id: ID;
  name: string;
  avatarUrl?: string;
}

export interface GeoLocation {
  name: string;
  lat?: number;
  lng?: number;
}

/**
 * AI-generated enrichment attached to a knowledge item. Populated by the
 * backend pipeline (transcription → summary → emotion → tags → ...). Optional
 * because enrichment may still be processing.
 */
export interface AIEnrichment {
  title?: string;
  summary?: string;
  mood?: Mood;
  topics?: string[];
  keyMoments?: string[];
  actionItems?: string[];
}

export interface KnowledgeItem extends HasTimestamps {
  id: ID;
  workspaceId: ID;
  collectionId?: ID;
  type: KnowledgeType;
  /** Full AI-pipeline state. The coarse status is derived from this. */
  pipeline: KnowledgePipeline;
  title: string;
  /** Raw text / transcript / body of the item. */
  content?: string;
  excerpt?: string;
  /** Small thumbnail for cards (image/video). */
  thumbnailUrl?: string;
  /** Renderable preview: the image itself, a PDF first-page render, a video poster. */
  previewUrl?: string;
  /** The original uploaded file, for download / open (object URL, then backend URL). */
  fileUrl?: string;
  /** Original source URL, for "Open source" (links/websites/bookmarks). */
  sourceUrl?: string;
  /** When the memory actually happened (may differ from createdAt). */
  occurredAt?: ISODateString;
  /** Last time the user opened this memory. */
  lastAccessedAt?: ISODateString;
  tags: string[];
  people: Person[];
  location?: GeoLocation;
  ai?: AIEnrichment;
  /**
   * RAG facets for AI Chat retrieval/citations. Optional — populated by the
   * backend; `deriveRag()` fills the gap until then.
   */
  rag?: KnowledgeRag;
  isFavorite: boolean;
}
